Importing the following libraries required for our program.


In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Numpy is to work with arrays, and pandas(python data analysis) is to work with data processing. The train_test_split function is used to split datasets/matrices/array.We import logisticregression model as thats what we will use for our program, due to the binary classification nature of our model. We import accuracy score to test the accuracy of our models.  We use the sklearn library as it is good for our machine learning algorithm.

Storing data and processing it as a panda dataframe

In [ ]:
# loading the data from the file into sonar data using pandas
Sd = pd.read_csv('Copy of sonar data.csv', header = None) # none as we do not have any header on our columns

In [ ]:
Sd.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

The rows represent an intance of scan, so at one time how the scan was. The column represents the different feature of the scan (at that one instance) so example of features could, be frequency and amplitude. The different rows are different instances of scan for different features and the last column is the label of it either being a rock or the mine (judges each instance differently based on readings of its features)

In [ ]:
Sd.shape # gets number of rows and columns

(208, 61)

In [ ]:
# obtaining statistical data about our dataset, gives statistical data for all the columns. Makes us better understand the data.
Sd.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [ ]:
Sd[60].value_counts() # tells us how many rock and mine example are there.We check for the 60th column as that is the column containing the rocks and mine

60
M    111
R     97
Name: count, dtype: int64

The data is well balanced for both rocks and mines as their count is similar and not far apart. M is mine and R is Rocks

In [ ]:
#Grouping the similar types together. This will give us the difference in mean of Mine and Rock for each column
Sd.groupby(60).mean()


0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

Spliting label from data to store label in a different variable

In [ ]:
#seperating label and features
X_NonLabel = Sd.drop(columns= 60)
X_Label = Sd[60]
print(X_NonLabel)
print(X_Label)

         0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53   

We now split noth sets of data X_NonLabel and X_Label into training and test set.

In [ ]:
# we split the data into train and test and gave those data their label accordingly, for both train and test data. Test size assigns the size of the test data as a percentage of the whole dataset, in this case 10% of data is test data.
# We use stratify to have equal number of rocks and mine in both test and training data
#Random state ensures that the data is split the same way everytime we run it, this helps with debugging and model performance. To ensure same split each time, we use random state.
X_NonLabel_Train,X_NonLabel_Test, X_Label_Train, X_Label_Test = train_test_split(X_NonLabel, X_Label, test_size = 0.1, stratify = X_Label, random_state = 1)

In [ ]:
#checking the amount of data in our train and test set
print(X_NonLabel.shape,X_NonLabel_Train.shape, X_NonLabel_Test.shape)
# we have 187 data to train and 21 data to test

(208, 60) (187, 60) (21, 60)


Now, we finally move to the logistic regression training of the model. We already imported the logistic regression model, we will train it with our training data.

In [ ]:
logistic_model = LogisticRegression()
# initialising the model
logistic_model.fit(X_NonLabel_Train, X_Label_Train)
#We train the model with X_NonLabel_Train, X_Label_Train


LogisticRegression()

In [ ]:
#Model Evaluation for training data, accuracy check. The model learnt on the training data, we will see how well it predicts that dataset
#Generally the accuracy is higher with more data points
X_NonLabel_Train_prediction = logistic_model.predict(X_NonLabel_Train) #we predict the label of our training data
training_data_accuracy = accuracy_score(X_Label_Train, X_NonLabel_Train_prediction) #we compare our model prediction with real label

In [ ]:
print("The training data accuracy score is:",training_data_accuracy )

The training data accuracy score is: 0.8342245989304813


Therefore, our training data accuracy score is 0.8342 which is 83.42%. This is pretty high and a good accuracy score, which is expected as our model learnt on the training data.

In [ ]:
#Accuracy score on our test data, how accurate is our test data
X_NonLabel_Test_prediction = logistic_model.predict(X_NonLabel_Test)
#Prediction of our test data by our logistic regression model
test_data_accuracy = accuracy_score(X_Label_Test, X_NonLabel_Test_prediction)
#comparing our prediction using our logistic model compared to the real label of that test data

In [ ]:
print("The test data accuracy score is:", test_data_accuracy)

The test data accuracy score is: 0.7619047619047619


We have an test accuracy of 76.2%, which is very good.

Constructing a predictive system, using our logistic model to predict if its a rock or mine.

In [ ]:
# Getting input from user, then splitting it by spaces, converting into float and hten to a numpy array using np
user_values = input("Enter the values seperated by either a space or a comma")
# we replace the commas with spaces and split the string by spaces
values_str = user_values.replace(',',' ').split()
# converting strings to float
values = [float(value) for value in values_str]
# we convert it into numpy array to process better
#converting input data list to numpy array
input_numpy = np.asarray(values)

Enter the values seperated by either a space or a comma0.0317	0.0956	0.1321	0.1408	0.1674	0.171	0.0731	0.1401	0.2083	0.3513	0.1786	0.0658	0.0513	0.3752	0.5419	0.544	0.515	0.4262	0.2024	0.4233	0.7723	0.9735	0.939	0.5559	0.5268	0.6826	0.5713	0.5429	0.2177	0.2149	0.5811	0.6323	0.2965	0.1873	0.2969	0.5163	0.6153	0.4283	0.5479	0.6133	0.5017	0.2377	0.1957	0.1749	0.1304	0.0597	0.1124	0.1047	0.0507	0.0159	0.0195	0.0201	0.0248	0.0131	0.007	0.0138	0.0092	0.0143	0.0036	0.0103


In [ ]:
# since we take input for one instance, we will reshape it for one instance. We use one instance to predict for that one instance
input_reshape = input_numpy.reshape(1,-1)
# 1, -1 represents one instance, we predict for that one instance

In [ ]:
# prediction for the instance
prediction = logistic_model.predict(input_reshape)
# our model is labelled as logistic_model and our data is input_reshape so we use them to predict for the data, input_reshape
print(prediction)

['R']
